### PR TITLE
Making some modifications to the way in which plans and protocols are…

### PR DESCRIPTION
**Pull request by cereallarceny**
Basically we're trying to switch the workers to a "multiple plans" and "multiple protocols" workflow, instead of being locked into one specific training plan and one optional protocol for the time being.

Currently, our list of unknowns and unsolved questions open for discussion are:

- [ ] How do we wire the outputs of a plan/protocol to the inputs of another plan, protocol, or report?

### CODE REVIEW

**Comment by vvmnnnkv** [comment]
> chain one "cycle" (or round) after another
> specified number of cycles / rounds or early stopping conditions (in plan / protocol?)
> all the workers are still in the same "session"

@Prtfw Hmm, do we need to chain cycles inside the Protocol? 
I think each cycle is a fresh start of everything, starting from the download of a new model checkpoint created by grid from the previous cycle. In a new cycle workers population can be different, so there will be no same session. 

> the baseworker class will implement something that will check if the received message "unblocks" some new computation... in this case it will know "I have received a result from a previous cycle" => "I still have x more cycles to go" => "I can start the next cycle"... the worker would know how many cycles it needs to run.

Hmm, maybe terminology issue. I think cycle is "workers ask pygrid for a job => workers get model checkpoint => workers train the model => (optionally) workers execute Protocol for sec agg => workers submit model diffs (or shares of avg model diffs) to pygrid => pygrid aggregates diffs into new model checkpoint". In this picture, worker doesn't know how many cycles are there and it doesn't decide to stop or continue, they just ask pygrid if there's a cycle to work on.
The early stopping logic can be programmed directly in pygrid and configured with params of `server_config`. But that's not in the roadmap currently.

> averaging plans are a lot like plans... what is the difference? is averaging plan just a function wrapped into a plan run by pygrid (which is basically like a worker?)
> averaging plans are a lot like protocols... what is the diff? it does some work and talks to other workers

Averaging plan is a function that tells pygrid how to take model diffs and update the model to get new checkpoint. It is serialized to Plan so it can be sent to pygrid. It's executed after the cycle is done and diffs are collected, so it doesn't talk to workers, and doesn't need to be Protocol. I can imagine it being a part of huge Protocol that controls whole cycle, but I'm not sure this is cleaner/easier that way, because such Protocol would need to have flexible roles count and be fault-tolerant.

NOTE: this is my understanding, which can be wrong!

File: web_and_mobile_team/projects/federated_learning.md [hunk 1]
```diff
@@ -253,9 +247,9 @@ Upon first glance, the example above is performing a lot of "magic" behind the s
...
 
-When the `on('ready')` event listener is triggered, this means that the worker has been chosen to participate by PyGrid and the model, training plan, and client configuration have been downloaded. Again, this magically happens in the background. At this point, the developer will likely batch their input and label data sets, and the worker may begin to run `executeTrainingPlan()`.
+When the `on('ready')` event listener is triggered, this means that the worker has been chosen to participate by PyGrid and the model, plan(s), protocol(s), and client configuration have been downloaded. Again, this magically happens in the background. At this point, the developer will likely batch their input and label data sets, and the worker may begin to run `executePlan()` against one of their plans.
```
Comment (vvmnnnkv) [inline web_and_mobile_team/projects/federated_learning.md]: You mention `executePlan()`, `executeProtocol()` here but in the code sample it's just `execute()`

**Comment by vvmnnnkv** [comment]
There's something here that doesn't feel comfortable...
We had specific training plan and secure aggregation protocol for a "training job" executed on worker.
The training job is not a generic "throw any plan at me" thing, it is specific kind of job that needs to get specific data, get specific model, run this training plan with this model and data as arguments, make diff, (optionally) encrypt diff using secure agg protocol, send diff for aggregation.
Now we try to make things more generic, but the job code is still "training job" that can't execute anything except training flow. So even if we put many plans and protocols into pygrid, worker would have no idea how to use them correctly. Moreover, it will be waste of bandwidth to download all those extra plans/protocols but not use them.

**Comment by vvmnnnkv** [comment]
Something comes to mind wrt wiring the outputs of plans/protocols and things mentioned above:
1. Have just one Protocol per job. If multiple plans or protocols needed, I think this can be always represented as bigger Protocol anyway? I.e. the inner wiring is done by Protocol. Input wiring (what should worker feed into Protocol) is still a question, e.g. we can add some metadata to Protocol so worker knows that this arg is data, this is arg is model, etc. Output wiring (from Protocol to report) is defined in Protocol.
2. Have a way to add multiple jobs to pygrid, e.g. model training job, model validation job, statistics collection job, etc.
3. Worker would ask for a job and may get any of them, whether pygrid decides in needs to execute right now. Worker would check if it can execute such job, i.e. if it understands all Protocol inputs and can supply them.


Pseudo code for training Protocol:
```
def train(model, data, target, optimizer):
    data, target = data
    optimizer.zero_grad()
    output = model(data)
    loss = F.nll_loss(output, target)
    loss.backward()
    optimizer.step()

worker1 = sy.VirtualWorker(hook=hook, id="worker1")
worker2 = sy.VirtualWorker(hook=hook, id="worker2")
worker3 = sy.VirtualWorker(hook=hook, id="worker3")

worker1_model = worker1.Promise.Model()
worker2_model = worker2.Promise.Model()
worker3_model = worker3.Promise.Model()

worker1_data = worker1.Promise.Data()
worker2_data = worker2.Promise.Data()
worker3_data = worker3.Promise.Data()

# Training plan inserted into Protocol
for _ in range(max_updates):
   train(worker1_model, worker1_data)
   train(worker2_model, worker2_data)
   train(worker3_model, worker3_data)

worker1_shares = worker1_model.share(worker1, worker2, worker3)
worker2_shares = worker2_model.share(worker1, worker2, worker3)
worker3_shares = worker3_model.share(worker1, worker2, worker3)

new_model_shares = (worker1_shares + worker2_shares + worker3_shares) / 3

new_model_shares["worker1"].declare_output()
new_model_shares["worker2"].declare_output()
new_model_shares["worker3"].declare_output()

training_protocol = sy.Protocol(worker1, worker2, worker3)
```

Pseudo code for worker:
```
const job = worker.newJob({
  model: 'my-federated-model'
});

job.start();
job.on('ready', async ({ config, protocol }) => {
  const args = await worker.resolve_inputs(config, protocol)
  const output = await protocol.execute(...args)
  job.report(output);
});

worker.resolve_inputs = (config, protocol) => { 
   let inputs = []
   for (let input of protocol.inputs) {
      switch (input.type) {
         case 'Data':
            inputs.push(rawData)
         case 'Model':
            inputs.push(worker.getModel(config.modelId))
         default:
            throw new Error(`I don't know what to do with this protocol!`)
      }
   }
  return inputs
}
```

**Comment by cereallarceny** [comment]
@Prtfw 

> as to how to chain one "cycle" (or round) after another, I think here are a way we can do it:

Don't worry about cycle chaining for now - we're getting a bit ahead of ourselves. This is not a problem to solve in this PR... a job for way down the line. 😄 

> averaging plans are a lot like plans... what is the difference? is averaging plan just a function wrapped into a plan run by pygrid (which is basically like a worker?)

An averaging plan is literally... just a plan CALLED the "averaging plan". It's called that because it's not a plan that the worker will run. It's a plan that PyGrid will run to... average the models and run the model update process. So, I think the confusion is just around naming. averaging plan === just another plan, but only for PyGrid. 👍 

> averaging plans are a lot like protocols... what is the diff? it does some work and talks to other workers

No, it's just a plan that's run by PyGrid to update the model. It has nothing to do with workers, other than the fact that it's averaging the model diffs submitted by the workers.

> if we can figure out the common parts, we can have a base class API and have wrappers / abstract classes that gives different behaviors and that will help the code get cleaner... I am not convinced we need a averaging_plan, a plan and a protocol class... i.e. it might be nice to have something like

See above.

> can a worker be a worker (doing some compute) for some other host, but also be a host to some other workers at the same time? (getting blockchain-ish I know...)

Yep. PyGrid is just a network of PySyft workers at the end of the day. You're maybe thinking of something close to EMLaaS, which will be approached in an upcoming grant. Basically a DS will approach an organization hosting a PyGrid network with data hosted on that network. They can submit their plans to the network to be trained on the data (on-premise, off the developer's machine). This transaction would be tracked via an epsilon value, which allows the network to monitor and meter such requests.

File: web_and_mobile_team/projects/federated_learning.md (hunk 1 above)
Comment (cereallarceny) [reply web_and_mobile_team/projects/federated_learning.md]: Resolved. Nice find @vvmnnnkv

**Comment by cereallarceny** [comment]
@vvmnnnkv 

> I think each cycle is a fresh start of everything, starting from the download of a new model checkpoint created by grid from the previous cycle. In a new cycle workers population can be different, so there will be no same session.

This is correct, you're on the right track here. A worker will not be able to continue their work onto a subsequent cycle unless the developer permits this in the `server_config`. This is currently referred to as the `do_not_reuse_workers_until_cycle: XXX` property where `XXX` is the number of cycles that must elapse until a worker is allowed to participate again.

> The early stopping logic can be programmed directly in pygrid and configured with params of server_config. But that's not described in detail in the roadmap currently.

Also correct. TBD

> Averaging plan is a function that tells pygrid how to take model diffs and update the model to get new checkpoint. It is serialized to Plan so it can be sent to pygrid. It's executed after the cycle is done and diffs are collected, so it doesn't talk to workers, and doesn't need to be Protocol. I can imagine it being a part of huge Protocol that controls whole cycle, but I'm not sure this is cleaner/easier that way, because such Protocol would need to have flexible roles count and be fault-tolerant.

Also correct.

> We had specific training plan and secure aggregation protocol for a "training job" executed on worker.
The training job is not a generic "throw any plan at me" thing, it is specific kind of job that needs to get specific data, get specific model, run this training plan with this model and data as arguments, make diff, (optionally) encrypt diff using secure agg protocol, send diff for aggregation.
Now we try to make things more generic, but the job code is still "training job" that can't execute anything except training flow. So even if we put many plans and protocols into pygrid, worker would have no idea how to use them correctly. Moreover, it will be waste of bandwidth to download all those extra plans/protocols but not use them.

The worker will know how to use them because the person implementing the worker is likely to be either the same person or working together with the model developer themselves. While we won't have one predefined "training plan" anymore, you can still create a "training plan"... you just need to create that plan and call it the "training plan". **If you're confused... just think that it's the exact same workflow as before, but now instead of saying `job.executeTrainingPlan(...myArgs)` you have to specify the plan you want to execute... like `job.plans['training_plan'].execute(...myArgs)`.**  It's the exact same thing, just written differently to allow for a bit more flexibility. We don't know if the model developer will want to execute multiple plans over multiple epochs before running the protocol or reporting back to PyGrid. This way they have the flexibility to decide how they want training to be done.

**Comment by cereallarceny** [comment]
@vvmnnnkv 

> Have just one Protocol per job. If multiple plans or protocols needed, I think this can be always represented as bigger Protocol anyway? I.e. the inner wiring is done by Protocol. Input wiring (what should worker feed into Protocol) is still a question, e.g. we can add some metadata to Protocol so worker knows that this arg is data, this is arg is model, etc. Output wiring (from Protocol to report) is defined in Protocol.

You're totally correct in your thinking here. I have no idea how we will be able to wire the outputs of a plan/protocol into the inputs of the next plan/protocol. This is the one open question for this PR and we need to solve it. Vova - you seem to have a good handle over this. Let's talk that conversation offline.

> Have a way to add multiple jobs to pygrid, e.g. model training job, model validation job, statistics collection job, etc.

I think this is somewhat built into the solution offered in this PR. If you want to... just spin up a second job and call it a day. 😄 

> Worker would ask for a job and may get any of them, whether pygrid decides in needs to execute right now. Worker would check if it can execute such job, i.e. if it understands all Protocol inputs and can supply them.

That's fine, but on the other hand... PyGrid can safely assume it won't hear back from all the workers and that's fine, too. Conditions change all the time: the user picks up their phone, the internet connection on a phone goes out, the phone comes off the charger, the worker doesn't have enough information to participate (even if they've been accepted into the cycle), etc. I don't think we need to build in special conditions for this... just have an understanding that sometimes a worker will not be able to continue.

**Comment by iamtrask** [comment]
I like @vvmnnnkv 's idea to simply make any multi-party protocol into a larger protocol. I think this is vastly superior to local plan/protocol chaining.

I'm also grateful that @vvmnnnkv contributed the idea while also saying that it shouldn't alter the immediate roadmap. Future planning while staying the course towards the MVP is great advice at this point.

More substantively, I should mention that there are a few assumptions with this - which would need to be included in any planning/refactor going in this direction. I think it's doable, but here they are:
1. we can't do IF statements in plans or protocols yet
2. we can't do FOR loops in plans or protocols yet - which means anything with variable length input across executions (aka... each worker has a different amount of training data) wouldn't necessarily work. Put another way - at the moment the only way to serialize a plan/protocol with a for loop is to liteerally duplicate the plan/protocol over and over and over a fixed number of times.
3. we would abandon the notion of compositional plans/protocols which could be re-used/re-composed. This was a nice idea primarily because we could pre-security vet some of them. This might be fixable if the "single protocol" was actually a nested set of "sub-protocols". We've done this with plans before, but doing it with protocols seems trickier.

All in all - a clever idea! Introduces a more more moles - but i think they're whackable :)

**Comment by cereallarceny** [comment]
Totally agree, we can go ahead and move forward with this PR as designated, but with the understanding that @vvmnnnkv's combined plan/protocol code will work in due time. The good news though is that this roadmap PR allows for both the current roadmap solution _right now_ as well as Vova's proposed solution _down the line_. I'm going to merge this. 😄

**Pull request merged**